File: apps/api-go/internal/events/events.go
package events

import (
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"

	"github.com/xirothedev/webdevstudios/apps/api-go/internal/cart"
)

type Event struct {
	ID          string    `gorm:"column:id"`
	Title       string    `gorm:"column:title"`
	Description *string   `gorm:"column:description"`
	StartDate   time.Time `gorm:"column:startDate"`
	EndDate     time.Time `gorm:"column:endDate"`
	Location    *string   `gorm:"column:location"`
	Type        string    `gorm:"column:type"`
	Organizer   *string   `gorm:"column:organizer"`
	Attendees   *int      `gorm:"column:attendees"`
	SurveyLink  *string   `gorm:"column:surveyLink"`
	CreatedBy   *string   `gorm:"column:createdBy"`
	CreatedAt   time.Time `gorm:"column:createdAt"`
	UpdatedAt   time.Time `gorm:"column:updatedAt"`
}

func (Event) TableName() string { return "events" }

type EventDTO struct {
	ID          string    `json:"id"`
	Title       string    `json:"title"`
	Description *string   `json:"description"`
	StartDate   time.Time `json:"startDate"`
	EndDate     time.Time `json:"endDate"`
	Location    *string   `json:"location"`
	Type        string    `json:"type"`
	Organizer   *string   `json:"organizer"`
	Attendees   *int      `json:"attendees"`
	SurveyLink  *string   `json:"surveyLink"`
	CreatedBy   *string   `json:"createdBy,omitempty"`
	CreatedAt   time.Time `json:"createdAt"`
	UpdatedAt   time.Time `json:"updatedAt"`
}

var validTypes = map[string]bool{"MEETING": true, "WORKSHOP": true, "SOCIAL": true, "COMPETITION": true, "SURVEY": true, "OTHER": true}

type Handler struct{ db *gorm.DB }

// Register mounts the whole module. Write routes are ADMIN-only like Nest's @Roles(ADMIN).
func Register(v1 *gin.RouterGroup, db *gorm.DB, authRequired, adminRequired gin.HandlerFunc) {
	h := &Handler{db: db}
	g := v1.Group("/events")
	g.GET("", h.list)
	g.GET("/:id", h.getByID)
	g.POST("", authRequired, adminRequired, h.create)
	g.PATCH("/:id", authRequired, adminRequired, h.update)
	g.DELETE("/:id", authRequired, adminRequired, h.delete)
}

func newID() string {
	b := make([]byte, 12)
	randRead(b)
	return hexEnc(b)
}

func (h *Handler) list(c *gin.Context) {
	q := h.db.Model(&Event{})
	if v := c.Query("startDate"); v != "" {
		if t, err := time.Parse(time.RFC3339, v); err == nil {
			q = q.Where("\"startDate\" >= ?", t)
		}
	}
	if v := c.Query("endDate"); v != "" {
		if t, err := time.Parse(time.RFC3339, v); err == nil {
			q = q.Where("\"endDate\" <= ?", t)
		}
	}
	if v := c.Query("types"); v != "" {
		types := splitCSV(v)
		q = q.Where("type IN ?", types)
	}
	var rows []Event
	if err := q.Order("\"startDate\" ASC").Find(&rows).Error; err != nil {
		fail(c, err)
		return
	}
	out := make([]EventDTO, 0, len(rows))
	for _, e := range rows {
		out = append(out, toDTO(e))
	}
	c.JSON(http.StatusOK, out)
}

func (h *Handler) getByID(c *gin.Context) {
	var e Event
	err := h.db.First(&e, map[string]any{"id": c.Param("id")}).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		nest(c, 404, fmt.Sprintf("Event with ID %s not found", c.Param("id")))
		return
	}
	if err != nil {
		fail(c, err)
		return
	}
	c.JSON(http.StatusOK, toDTO(e))
}

func (h *Handler) create(c *gin.Context) {
	var in struct {
		Title       string  `json:"title" binding:"required,max=255"`
		Description *string `json:"description"`
		StartDate   string  `json:"startDate" binding:"required"`
		EndDate     string  `json:"endDate" binding:"required"`
		Location    *string `json:"location"`
		Type        string  `json:"type" binding:"required"`
		Organizer   *string `json:"organizer"`
		Attendees   *int    `json:"attendees"`
		SurveyLink  *string `json:"surveyLink"`
	}
	if err := c.ShouldBindJSON(&in); err != nil {
		nest(c, http.StatusBadRequest, err.Error())
		return
	}
	if !validTypes[in.Type] {
		nest(c, http.StatusBadRequest, "type must be one of MEETING, WORKSHOP, SOCIAL, COMPETITION, SURVEY, OTHER")
		return
	}
	sd, err1 := time.Parse(time.RFC3339, in.StartDate)
	ed, err2 := time.Parse(time.RFC3339, in.EndDate)
	if err1 != nil || err2 != nil {
		nest(c, http.StatusBadRequest, "startDate/endDate must be ISO date strings")
		return
	}
	e := Event{
		ID: newID(), Title: in.Title, Description: in.Description,
		StartDate: sd, EndDate: ed, Location: in.Location, Type: in.Type,
		Organizer: in.Organizer, Attendees: in.Attendees, SurveyLink: in.SurveyLink,
		CreatedBy: strPtr(c.GetString("userId")),
	}
	if err := h.db.Create(&e).Error; err != nil {
		fail(c, err)
		return
	}
	c.JSON(http.StatusCreated, toDTO(e))
}

func (h *Handler) update(c *gin.Context) {
	var e Event
	err := h.db.First(&e, map[string]any{"id": c.Param("id")}).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		nest(c, 404, fmt.Sprintf("Event with ID %s not found", c.Param("id")))
		return
	}
	var in map[string]any
	if err := c.ShouldBindJSON(&in); err != nil {
		nest(c, http.StatusBadRequest, err.Error())
		return
	}
	updates := map[string]any{}
	for _, k := range []string{"title", "description", "location", "organizer", "surveyLink"} {
		if v, ok := in[k]; ok {
			updates[k] = v
		}
	}
	if v, ok := in["type"]; ok {
		s, _ := v.(string)
		if !validTypes[s] {
			nest(c, http.StatusBadRequest, "invalid event type")
			return
		}
		updates["type"] = s
	}
	if v, ok := in["attendees"]; ok {
		f, _ := v.(float64)
		i := int(f)
		updates["attendees"] = i
	}
	for _, k := range []string{"startDate", "endDate"} {
		if v, ok := in[k]; ok {
			s, _ := v.(string)
			t, err := time.Parse(time.RFC3339, s)
			if err != nil {
				nest(c, http.StatusBadRequest, k+" must be an ISO date string")
				return
			}
			updates[k] = t
		}
	}
	if len(updates) > 0 {
		if err := h.db.Model(&Event{}).Where(map[string]any{"id": e.ID}).Updates(updates).Error; err != nil {
			fail(c, err)
			return
		}
	}
	h.db.First(&e, map[string]any{"id": e.ID})
	c.JSON(http.StatusOK, toDTO(e))
}

func (h *Handler) delete(c *gin.Context) {
	res := h.db.Delete(&Event{}, map[string]any{"id": c.Param("id")})
	if res.Error != nil {
		fail(c, res.Error)
		return
	}
	if res.RowsAffected == 0 {
		nest(c, 404, fmt.Sprintf("Event with ID %s not found", c.Param("id")))
		return
	}
	c.JSON(http.StatusOK, gin.H{"success": true})
}

func toDTO(e Event) EventDTO {
	return EventDTO{
		ID: e.ID, Title: e.Title, Description: e.Description,
		StartDate: e.StartDate, EndDate: e.EndDate, Location: e.Location,
		Type: e.Type, Organizer: e.Organizer, Attendees: e.Attendees,
		SurveyLink: e.SurveyLink, CreatedBy: e.CreatedBy,
		CreatedAt: e.CreatedAt, UpdatedAt: e.UpdatedAt,
	}
}

func fail(c *gin.Context, err error) {
	var he *cart.HTTPError
	if errors.As(err, &he) {
		nest(c, he.Code, he.Msg)
		return
	}
	nest(c, http.StatusInternalServerError, err.Error())
}

func nest(c *gin.Context, code int, msg string) {
	c.JSON(code, gin.H{"statusCode": code, "message": msg, "error": http.StatusText(code)})
}

func splitCSV(s string) []string {
	var out []string
	cur := ""
	for _, r := range s {
		if r == ',' {
			if cur != "" {
				out = append(out, cur)
			}
			cur = ""
			continue
		}
		cur += string(r)
	}
	if cur != "" {
		out = append(out, cur)
	}
	return out
}

func strPtr(s string) *string {
	if s == "" {
		return nil
	}
	return &s
}

func randRead(b []byte) { _, _ = rand.Read(b) }

func hexEnc(b []byte) string { return hex.EncodeToString(b) }

File: apps/api-elysia/src/routes/events.ts
import { Elysia } from 'elysia';
import type { Event, EventType } from '../generated/prisma/client';
import { ApiError } from '../lib/errors';
import { db } from '../lib/prisma';
import { bindBody, readJsonObject } from '../lib/validate';
import { requireAdmin } from '../lib/auth';
import { goTime, newId } from '../lib/util';

// Mirrors events.validTypes
const VALID_TYPES = ['MEETING', 'WORKSHOP', 'SOCIAL', 'COMPETITION', 'SURVEY', 'OTHER'];

const ISO_RE = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}(:\d{2}(\.\d+)?)?(Z|[+-]\d{2}:\d{2})$/;

// ponytail: Go's time.Parse(time.RFC3339) is strict; regex + Date.parse is a close approximation.
function parseIso(v: unknown): Date | null {
  if (typeof v !== 'string' || !ISO_RE.test(v)) return null;
  const t = Date.parse(v);
  return Number.isNaN(t) ? null : new Date(t);
}

function toDTO(e: Event) {
  const out: Record<string, unknown> = {
    id: e.id,
    title: e.title,
    description: e.description,
    startDate: goTime(e.startDate),
    endDate: goTime(e.endDate),
    location: e.location,
    type: e.type,
    organizer: e.organizer,
    attendees: e.attendees,
    surveyLink: e.surveyLink,
    createdAt: goTime(e.createdAt),
    updatedAt: goTime(e.updatedAt),
  };
  if (e.createdBy !== null) out.createdBy = e.createdBy;
  return out;
}

export const events = new Elysia()
  .get('/events', async ({ query }) => {
    const where: Record<string, unknown> = {};
    const start = parseIso(query.startDate);
    if (start) where.startDate = { gte: start };
    const end = parseIso(query.endDate);
    if (end) where.endDate = { lte: end };
    if (query.types) {
      const types = String(query.types)
        .split(',')
        .filter((t) => t !== '');
      if (types.length > 0) where.type = { in: types };
    }
    const rows = await db().event.findMany({ where, orderBy: { startDate: 'asc' } });
    return rows.map(toDTO);
  })
  .get('/events/:id', async ({ params }) => {
    const e = await db().event.findUnique({ where: { id: params.id! } });
    if (e === null) throw new ApiError(404, `Event with ID ${params.id} not found`);
    return toDTO(e);
  })
  .post('/events', async ({ request, cookie, body, set }) => {
    await requireAdmin({ request, cookie });
    const in1 = bindBody<{
      title?: string;
      description?: string | null;
      startDate?: string;
      endDate?: string;
      location?: string | null;
      type?: string;
      organizer?: string | null;
      attendees?: number | null;
      surveyLink?: string | null;
    }>(body as Record<string, unknown>, {
      Title: { type: 'string', required: true, maxLen: 255 },
      StartDate: { type: 'string', required: true },
      EndDate: { type: 'string', required: true },
      Type: { type: 'string', required: true },
      Description: { type: 'string' },
      Location: { type: 'string' },
      Organizer: { type: 'string' },
      Attendees: { type: 'number', integer: true },
      SurveyLink: { type: 'string' },
    });
    if (!VALID_TYPES.includes(in1.type ?? '')) {
      throw new ApiError(
        400,
        'type must be one of MEETING, WORKSHOP, SOCIAL, COMPETITION, SURVEY, OTHER',
      );
    }
    const sd = parseIso(in1.startDate);
    const ed = parseIso(in1.endDate);
    if (sd === null || ed === null) {
      throw new ApiError(400, 'startDate/endDate must be ISO date strings');
    }
    const e = await db().event.create({
      data: {
        id: newId(),
        title: in1.title!,
        description: in1.description ?? null,
        startDate: sd,
        endDate: ed,
        location: in1.location ?? null,
        type: in1.type! as EventType,
        organizer: in1.organizer ?? null,
        attendees: in1.attendees ?? null,
        surveyLink: in1.surveyLink ?? null,
      },
    });
    set.status = 201;
    return toDTO(e);
  })
  .patch('/events/:id', async ({ request, cookie, params }) => {
    await requireAdmin({ request, cookie });
    const in1 = await readJsonObject(request);
    const e = await db().event.findUnique({ where: { id: params.id! } });
    if (e === null) throw new ApiError(404, `Event with ID ${params.id} not found`);
    const updates: Record<string, unknown> = {};
    for (const k of ['title', 'description', 'location', 'organizer', 'surveyLink']) {
      const v = in1[k];
      if (typeof v === 'string') updates[k] = v;
    }
    if ('type' in in1) {
      const s = typeof in1.type === 'string' ? in1.type : '';
      if (!VALID_TYPES.includes(s)) throw new ApiError(400, 'invalid event type');
      updates.type = s;
    }
    if ('attendees' in in1) {
      // ponytail: mirrors Go `f, _ := v.(float64); int(f)` — non-number zero-fills.
      const f = typeof in1.attendees === 'number' ? in1.attendees : 0;
      updates.attendees = Math.trunc(f);
    }
    for (const k of ['startDate', 'endDate']) {
      if (k in in1) {
        const t = parseIso(in1[k]);
        if (t === null) throw new ApiError(400, `${k} must be an ISO date string`);
        updates[k] = t;
      }
    }
    if (Object.keys(updates).length > 0) {
      updates.updatedAt = new Date();
      const updated = await db().event.update({ where: { id: e.id }, data: updates });
      return toDTO(updated);
    }
    return toDTO(e);
  })
  .delete('/events/:id', async ({ request, cookie, params }) => {
    await requireAdmin({ request, cookie });
    const res = await db().event.deleteMany({ where: { id: params.id! } });
    if (res.count === 0) throw new ApiError(404, `Event with ID ${params.id} not found`);
    return { success: true };
  });

export default events;
